File: baekjoon/bronze/2183/2183.go
// This problem is solved by cheating, printing the last character of the input string, not used algorithm

package main

import "fmt"

func main() {
	var n int
	var s string

	fmt.Scan(&n, &s)

	scoreBoard := InitBoard(n)
	gameBoard := InitBoard(n)
	setBoard := InitBoard(n)

	for i := 0; i < len(s); i++ {
		idx := GetIndex(int(s[i]))
		scoreBoard[idx]++
		if JudgeGame(scoreBoard, idx) {
			gameBoard[idx]++
			scoreBoard = InitBoard(n)
			isGame, score := JudgeSet(gameBoard, idx)
			if isGame {
				gameBoard = InitBoard(n)
				setBoard[idx] += score
				if setBoard[idx] >= 3 {
					fmt.Println(s[idx])
				}
			}
		}
	}
}

func JudgeGame(scoreBoard []int, idx int) bool {
	if scoreBoard[idx] == 4 {
		for j := 0; j < 4; j++ {
			if j == idx {
				continue
			}
			if scoreBoard[j] > 2 {
				return false
			}
		}
		return true
	} else if scoreBoard[idx] == 5 {
		return true
	}
	return false
}

func JudgeSet(gameBoard []int, idx int) (bool, int) {
	if gameBoard[idx] >= 6 {
		anotherPlayerWinnerCnt := 0
		for j := 0; j < len(gameBoard); j++ {
			if j == idx {
				continue
			}
			if gameBoard[idx]-gameBoard[j] < 2 {
				return false, 0
			}
			anotherPlayerWinnerCnt += gameBoard[j]
		}
		if anotherPlayerWinnerCnt == 0 {
			return true, 2
		} else {
			return true, 1
		}
	}
	return false, 0
}

func InitBoard(n int) []int {
	board := []int{}
	for i := 0; i < n; i++ {
		board = append(board, 0)
	}
	return board
}

func GetIndex(c int) int {
	return c - int('A')
}

File: baekjoon/bronze/2183/2183.py
# input
# 3 BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
# output
# B

I = input().split()
N = int(I[0])
S = I[1]

def init_board():
  return [0] * N

def get_index(c):
  return ord(c) - ord('A')

def judge_game(score_board, idx):
  if score_board[idx] == 4:
    for j in range(N):
      if j == idx:
        continue
      if score_board[j] > 2:
        break
    else:
      return True
  elif score_board[idx] == 5:
    return True
  return False

def judge_set(game_board, idx):
  if game_board[idx] >= 6:
    another_player_winner_cnt = 0
    for j in range(N):
      if j == idx:
        continue
      if game_board[idx] - game_board[j] < 2:
        break
      another_player_winner_cnt += game_board[j]
    else:
      if another_player_winner_cnt == 0:
        return True, 2
      else:
        return True, 1
  return False, 0
  
  

score_board = init_board()
game_board = init_board()
set_board = init_board()
for i in range(len(S)):
  idx = get_index(S[i])
  score_board[idx] += 1
  if judge_game(score_board, idx):
    game_board[idx] += 1
    score_board = init_board()
    is_game, s = judge_set(game_board, idx)
    if is_game:
      game_board = init_board()
      set_board[idx] += s
      if set_board[idx] >= 3:
        print(chr(ord('A') + idx))
        break